### ISSUE TITLE
godaddy: test suite is failing, delete throws HTTP 422

### ISSUE THREAD

**Issue by alexzorin** [AnalogJ/lexicon#214]
- [Reported on the Let's Encrypt forums](https://community.letsencrypt.org/t/getting-unittest-2-1-1-0-requires-argparse-error/58625/6).
- Test suite fails against provided OTE endpoint and credentials: `tests/providers/test_godaddy.py .......ssFFFsssFsssFFsF.F.FF`
- Confirmed by me against a live domain with live credentials, fails in same way as forum post:

```
Arguments: Namespace(action='delete', auth_key='<redacted>', auth_secret='<redacted>', content='abcdef', delegated=None, domain='monkas.gq', identifier=None, log_level='DEBUG', name='_acme-challenge.monkas.gq', priority=None, provider_name='godaddy', ttl=None, type='TXT')
Starting new HTTPS connection (1): api.godaddy.com
https://api.godaddy.com:443 "GET /v1/domains/monkas.gq HTTP/1.1" 200 212
Starting new HTTPS connection (1): api.godaddy.com
https://api.godaddy.com:443 "GET /v1/domains/monkas.gq/records HTTP/1.1" 200 251
Starting new HTTPS connection (1): api.godaddy.com
https://api.godaddy.com:443 "PUT /v1/domains/monkas.gq/records HTTP/1.1" 422 275
Traceback (most recent call last):
  File "/home/alex/devel/venv-lexicon/bin/lexicon", line 11, in <module>
    sys.exit(main())
  File "/home/alex/devel/venv-lexicon/local/lib/python2.7/site-packages/lexicon/__main__.py", line 72, in main
    client.execute()
  File "/home/alex/devel/venv-lexicon/local/lib/python2.7/site-packages/lexicon/client.py", line 50, in execute
    return self.provider.delete_record(self.options.get('identifier'), self.options.get('type'), self.options.get('name'), self.options.get('content'))
  File "/home/alex/devel/venv-lexicon/local/lib/python2.7/site-packages/lexicon/providers/godaddy.py", line 106, in delete_record
    self._put('/domains/{0}/records'.format(domain), to_insert)
  File "/home/alex/devel/venv-lexicon/local/lib/python2.7/site-packages/lexicon/providers/base.py", line 79, in _put
    return self._request('PUT', url, data=data, query_params=query_params)
  File "/home/alex/devel/venv-lexicon/local/lib/python2.7/site-packages/lexicon/providers/godaddy.py", line 130, in _request
    result.raise_for_status()
  File "/home/alex/devel/venv-lexicon/local/lib/python2.7/site-packages/requests/models.py", line 935, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 422 Client Error: Unprocessable Entity for url: https://api.godaddy.com/v1/domains/monkas.gq/records
```

**Comment by alexzorin** [comment]
I think the recordings are fine, but the Live API is failing.

I don't really understand the GoDaddy API nor do I use their services, so I'm not keen to investigate this myself, just cross-posting the issue from the forum.

GoDaddy DNS hosting is usable from a free account (though they delete your zone after some time if it's not attached to a domain).

cc @adferrand

**Comment by jvanasco** [comment]
This may linked to the issue i brought up in #209.  In that ticket what I noticed is:

1. adding a new test to core requires new recordings
2. the test system hits the live API with invalid credentials
3. godaddy pushes the caller's IP into a blacklist which both forbids and tarpits (rate-limits) at increasing distances (the response takes 2x as long after each attempt).

I think once you fail against their API a few times (perhaps 3?) you need to wait an unspecified time before you can make a successful attempt from your IP address. I could be wrong, as I'm just reverse engineering their behavior... but this is what seems to happen.

**Comment by tubit** [comment]
I just tried to create a TXT record on my domain at godaddy, this is the command output:
```
# lexicon godaddy create konert.me TXT --auth-key AUTH_KEY --auth-secret AUTH_SECRET --name '_acme-challenge
.konert.me' --content ACME_CONTENT
Arguments: Namespace(action='create', auth_key='AUTH_KEY', auth_secret='AUTH_KEY', content='ACME_CONTENT', delegated=None, domain='konert.me', identifier=None, log_level='DEBUG', name='_acme-challenge.konert.me', priority=None, provider_name='godaddy', ttl=None, type='TXT')
Starting new HTTPS connection (1): api.godaddy.com
https://api.godaddy.com:443 "GET /v1/domains/konert.me HTTP/1.1" 200 438
Starting new HTTPS connection (1): api.godaddy.com
https://api.godaddy.com:443 "PUT /v1/domains/konert.me/records/TXT/_acme-challenge HTTP/1.1" 200 0
Traceback (most recent call last):
  File "/usr/local/bin/lexicon", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.6/site-packages/lexicon/__main__.py", line 72, in main
    client.execute()
  File "/usr/local/lib/python3.6/site-packages/lexicon/client.py", line 41, in execute
    return self.provider.create_record(self.options.get('type'), self.options.get('name'), self.options.get('content'))
  File "/usr/local/lib/python3.6/site-packages/lexicon/providers/godaddy.py", line 35, in create_record
    .format(domain, type, self._relative_name(name)), [data])
  File "/usr/local/lib/python3.6/site-packages/lexicon/providers/base.py", line 79, in _put
    return self._request('PUT', url, data=data, query_params=query_params)
  File "/usr/local/lib/python3.6/site-packages/lexicon/providers/godaddy.py", line 131, in _request
    return result.json()
  File "/usr/local/lib/python3.6/site-packages/requests/models.py", line 892, in json
    return complexjson.loads(self.text, **kwargs)
  File "/usr/local/lib/python3.6/json/__init__.py", line 354, in loads
    return _default_decoder.decode(s)
  File "/usr/local/lib/python3.6/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/local/lib/python3.6/json/decoder.py", line 357, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

But however, the TXT record was created:
```
# host -t TXT _acme-challenge.konert.me
_acme-challenge.konert.me descriptive text "ACME_CHALLENGE"
```

If I try to delete the created entry, I got the following output:

```
# lexicon godaddy delete konert.me TXT --auth-key AUTH_KEY --auth-secret AUTH_SECRET --name '_acme-challenge
.konert.me' --content 'ACME_CHALLENGE'
Arguments: Namespace(action='delete', auth_key='AUTH_KEY', auth_secret='AUTH_SECRET', content='ACME_CHALLENGE', delegated=None, domain='konert.me', identifier=None, log_level='DEBUG', name='_acme-challenge.konert.me', priority=None, provider_name='godaddy', ttl=None, type='TXT')
Starting new HTTPS connection (1): api.godaddy.com
https://api.godaddy.com:443 "GET /v1/domains/konert.me HTTP/1.1" 200 438
Starting new HTTPS connection (1): api.godaddy.com
https://api.godaddy.com:443 "GET /v1/domains/konert.me/records HTTP/1.1" 200 429
Starting new HTTPS connection (1): api.godaddy.com
https://api.godaddy.com:443 "PUT /v1/domains/konert.me/records HTTP/1.1" 422 985
Traceback (most recent call last):
  File "/usr/local/bin/lexicon", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.6/site-packages/lexicon/__main__.py", line 72, in main
    client.execute()
  File "/usr/local/lib/python3.6/site-packages/lexicon/client.py", line 50, in execute
    return self.provider.delete_record(self.options.get('identifier'), self.options.get('type'), self.options.get('name'), self.options.get('content'))
  File "/usr/local/lib/python3.6/site-packages/lexicon/providers/godaddy.py", line 106, in delete_record
    self._put('/domains/{0}/records'.format(domain), to_insert)
  File "/usr/local/lib/python3.6/site-packages/lexicon/providers/base.py", line 79, in _put
    return self._request('PUT', url, data=data, query_params=query_params)
  File "/usr/local/lib/python3.6/site-packages/lexicon/providers/godaddy.py", line 130, in _request
    result.raise_for_status()
  File "/usr/local/lib/python3.6/site-packages/requests/models.py", line 935, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 422 Client Error: Unprocessable Entity for url: https://api.godaddy.com/v1/domains/konert.me/records
```

Of cause the TXT record was deleted despite the error message.

```
# lexicon --version
lexicon 2.2.1
```

**Comment by alexzorin** [comment]
I've done a bit more digging and it looks like the provider is kinda broken.

- GoDaddy API returns empty response body for many of its `PUT` endpoints, but provider assumes a JSON response body, which will always result in a raised error when reading the response.
- Does not support creation of multiple records of the same type (e.g. multiple TXT records just overwrite each other)
- Deletion appears to not work at all, the sent request is not what the API expects
- (not an exhaustive list)

@AnalogJ since the codeowner (@adferrand) hasn't responded, should it just be deleted/disabled until it can be fixed? At least the README should not advertise that it is supported.

**Comment by adferrand** [comment]
Hello, I would like to repair it, but I could not find the time to do it. I'll try to do as soon as I can, but I cannot guarantee a deadline for now.

**Comment by adferrand** [comment]
By the way, I implemented the provider using the demo, which was by this time a quite difficult to use with a lot of restrictions. @alexzorin, would you like to give me an access to one of your domain for the time of the re-implementation ? It would help me a lot.

**Comment by alexzorin** [comment]
Thanks @adferrand . I don't have a "real" domain there, I just set one up using a free GoDaddy account and adding a free domain from freenom.com. GoDaddy seems to disable domains not attached to a paid service every 30 days, though.

If you want to avoid signing up I can share my API key and secret with you with my test account and couple of freenom domains, just email me via my GitHub profile.

**Comment by netcookies** [comment]
+1 for same problems.

**Comment by adferrand** [comment]
OK I took the time to work on godaddy provider implementation. I think I haved fixed all issues, PR will be submitted tomorrow.

**Comment by alexzorin** [comment]
Appears to be fixed by #241 , thanks!
